File: interviews/lc/05-single-number.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
#include "lc-utility.h"

using std::string;
using std::unordered_map;
using std::unordered_set;
using std::vector;
using std::chrono::high_resolution_clock;

int singleNumberHash(const vector<int>& nums) {
  unordered_map<int, size_t> numberFreqs;
  for (const auto n : nums) {
    numberFreqs[n]++;
  }
  for (const auto& p : numberFreqs) {
    if (p.second == 1) {
      return p.first;
    }
  }
  assert(false && "Unreachable");
}

int singleNumberMath(const vector<int>& nums) {
  int sumSet = 0;
  unordered_set<int> numSet(nums.begin(), nums.end());
  for (const auto n : numSet) {
    sumSet += n;
  }
  int sum = 0;
  for (const auto n : nums) {
    sum += n;
  }
  return 2 * sumSet - sum;
}

int singleNumberXor(const vector<int>& nums) {
  int result = 0;
  for (const auto n : nums) {
    result ^= n;
  }
  return result;
}

void runTestCase(const vector<int>& nums, int expected) {
  const auto numsStr = vector2str(nums);
  {
    const auto start = high_resolution_clock::now();
    const auto result = singleNumberHash(nums);
    const auto end = high_resolution_clock::now();
    std::cout << "singleNumberHash: " << numsStr << " Actual: " << result
              << " Expected: " << expected << " "
              << std::chrono::nanoseconds(end - start).count() << "ns"
              << std::endl;
  }
  {
    const auto start = high_resolution_clock::now();
    const auto result = singleNumberMath(nums);
    const auto end = high_resolution_clock::now();
    std::cout << "singleNumberMath: " << numsStr << " Actual: " << result
              << " Expected: " << expected << " "
              << std::chrono::nanoseconds(end - start).count() << "ns"
              << std::endl;
  }
  {
    const auto start = high_resolution_clock::now();
    const auto result = singleNumberXor(nums);
    const auto end = high_resolution_clock::now();
    std::cout << "singleNumberXor: " << numsStr << " Actual: " << result
              << " Expected: " << expected << " "
              << std::chrono::nanoseconds(end - start).count() << "ns"
              << std::endl;
  }
}

int main(int argc, char* argv[]) {
  runTestCase({2, 2, 1}, 1);
  runTestCase({4, 1, 2, 1, 2}, 4);
  runTestCase({1}, 1);
  runTestCase({1, 2, 3, 1, 2, 3, 4}, 4);
  return EXIT_SUCCESS;
}

File: interviews/lc/05-single-number.py
from typing import List
import time


def singleNumberHash(nums: List[int]) -> int:
    freqs = {}
    for i in nums:
        if i in freqs:
            freqs[i] += 1
        else:
            freqs[i] = 1
    for k, v in freqs.items():
        if v == 1:
            return k
    assert False, "Unreachable"


def singleNumberMath(nums: List[int]) -> int:
    """
    2 * (a + b + c) - (a + a + b + b + c) = c
    """
    return 2 * sum(set(nums)) - sum(nums)


def singleNumberXor(nums: List[int]) -> int:
    result = 0
    for n in nums:
        result ^= n
    return result


def runTestCase(nums: List[int], expected: int) -> None:
    numsStr = ",".join([str(n) for n in nums])
    start = time.perf_counter_ns()
    result = singleNumberHash(nums)
    end = time.perf_counter_ns()
    delta = end - start
    print(f"{numsStr} => Actual: {result}, Expected: {expected} {delta}ns")

    start = time.perf_counter_ns()
    result = singleNumberMath(nums)
    end = time.perf_counter_ns()
    delta = end - start
    print(f"{numsStr} => Actual: {result}, Expected: {expected} {delta}ns")

    start = time.perf_counter_ns()
    result = singleNumberXor(nums)
    end = time.perf_counter_ns()
    delta = end - start
    print(f"{numsStr} => Actual: {result}, Expected: {expected} {delta}ns")


if __name__ == "__main__":
    runTestCase([2, 2, 1], 1)
    runTestCase([4, 1, 2, 1, 2], 4)
    runTestCase([1], 1)
    runTestCase([1, 2, 3, 1, 2, 3, 4], 4)
